src/frontend/servers/y-provider/src/server.ts

```diff
@@ -1,18 +1,142 @@
+import expressWebsockets from 'express-ws';
 
-const port = Number(process.env.PORT || 4444);
+import { PORT } from './env';
+import { httpSecurity, wsSecurity } from './middlelayers';
+import { routes } from './routes';
+import { logger } from './utils';
 
-const server = Server.configure({
-  name: 'docs-y-provider',
-  port: port,
+export const hocuspocusServer = Server.configure({
+  name: 'docs-y-server',
   timeout: 30000,
-  debounce: 2000,
-  maxDebounce: 30000,
   quiet: true,
-});
+  onConnect({ requestHeaders, connection, documentName, requestParameters }) {
+    const roomParam = requestParameters.get('room');
+    const canEdit = requestHeaders['x-can-edit'] === 'True' ? true : false;
+
+    if (!canEdit) {
+      connection.readOnly = true;
+    }
+
+    logger(
+      'Connection established:',
+      documentName,
+      'userId:',
+      requestHeaders['x-user-id'],
+      'canEdit:',
+      canEdit,
+      'room:',
+      requestParameters.get('room'),
+    );
+
+    if (documentName !== roomParam) {
+      console.error(
+        'Invalid room name - Probable hacking attempt:',
+        documentName,
+        requestParameters.get('room'),
+        requestHeaders['x-user-id'],
+      );
+
+      return Promise.reject(new Error('Unauthorized'));
+    }
 
-server.listen().catch((error) => {
-  console.error('Failed to start the server:', error);
+    return Promise.resolve();
+  },
 });
 
-console.log('Websocket server running on port :', port);
+/**
+ * init the collaboration server.
+ *
+ * @param port - The port on which the server listens.
+ * @param serverSecret - The secret key for API authentication.
+ * @returns An object containing the Express app, Hocuspocus server, and HTTP server instance.
+ */
+export const initServer = () => {
+  const { app } = expressWebsockets(express());
+  app.use(express.json());
+
+  /**
+   * Route to handle WebSocket connections
+   */
+  app.ws(routes.WS, wsSecurity, (ws, req) => {
+    logger('Incoming Origin:', req.headers['origin']);
+
+    try {
+      hocuspocusServer.handleConnection(ws, req);
+    } catch (error) {
+      console.error('Failed to handle WebSocket connection:', error);
+      ws.close();
+    }
+  });
+
+  type ResetConnectionsRequestQuery = {
+    room?: string;
+  };
+
+  /**
+   * Route to reset connections in a room
+   */
+  app.post(
+    routes.RESET_CONNECTIONS,
+    httpSecurity,
+    async (
+      req: Request<object, object, object, ResetConnectionsRequestQuery>,
+      res: Response,
+    ) => {
+      const room = req.query.room;
+      const userId = req.headers['x-user-id'];
+
+      logger(
+        'Resetting connections in room:',
+        room,
+        'for user:',
+        userId,
+        'room:',
+        room,
+      );
+
+      if (!room) {
+        res.status(400).json({ error: 'Room name not provided' });
+        return;
+      }
+
+      const docConnection = await hocuspocusServer.openDirectConnection(room);
```

Comment:
I think you don't need to call `openDirectConnection`, but it you can just access `hocuspocusServer.documents`

Reply:
Actually, there seems to be a `hocuspocusServer.closeConnections(docName)`

Reply:
Improved:
https://github.com/numerique-gouv/impress/pull/472/commits/096837a123543a7cda9396168e3e0ae37bdf54bf#diff-aabe2f42c0dee383e91e6ad77a11be21ddb2246bee1007ee1678dcc09c55557e